### PR TITLE
Handle nil versions in preparation for rubygems 4

### LINKED ISSUE #11744: nil versions are discouraged and will be deprecated in Rubygems 4
`nil versions are discouraged and will be deprecated in Rubygems 4`

Potentially effects every module which makes use of `Gem::Version.new`.

### PR DESCRIPTION
**Pull request by dwelch-r7**
Resolves #11744 

Some information about nil being deprecated in an upcoming Rubygems version here: https://github.com/rubygems/rubygems/issues/2359

This PR wraps the `Gem::Version` from Rubygems and handles a `nil` version being passed in in such a way as to keep the current functionality (treats it equal to the version being 0)

Vast majority of changes are just changing references from `Gem::Version` to `Rex::Version`, important files are here 
`lib/rex/version.rb` and the tests here `spec/lib/rex/version_spec.rb`

# Verification
- [x] Tests pass

### CODE REVIEW

**Comment by gwillcox-r7** [comment]
Rebasing due to merge conflicts. Should hopefully be resolved now, the issues seemed to be pretty minor, only thing that may break is some minor RuboCop checks that were applied which I tried reapplying as best I could. Can always run these checks again though

File: lib/rex.rb
Comment (adfoster-r7) [inline lib/rex.rb:100]: @gwillcox-r7 Looks like some unrelated code has snuck in from the rebase 👀 

If you're happy waiting I think Dean will be able to fix this up on Monday :+1:

**Review by adfoster-r7 — CHANGES_REQUESTED** [review]

File: lib/rex.rb
Comment (gwillcox-r7) [reply lib/rex.rb:100]: Ok no problem

File: lib/rex/version.rb
Comment (adfoster-r7) [inline lib/rex/version.rb:12]: 💎 👮‍♂️ 👀

Comment (dwelch-r7) [reply lib/rex/version.rb:12]: I'll rebase this and give it a once over with rubocop soon after I'm done with my current task

File: lib/rex/version.rb
Comment (adfoster-r7) [inline lib/rex/version.rb:8]: I think this approach makes sense to me for now, it keeps us consistent with the current API. Long term we should decide how we want to the scenario of handle nil/empty strings. Nothing actionable for this PR though.

File: lib/rex/version.rb
Comment (dwelch-r7) [reply lib/rex/version.rb:11]: done :)

**Comment by gwillcox-r7** [comment]
Looks like the linter is failing on a few modules unrelated to this due to #14735. Let me go ahead and make a new PR to fix up those issues so the linter doesn't keep failing on new modules.

**Comment by gwillcox-r7** [comment]
Marking this as tests passed as whilst msftidy linter is failing, this appears to be unrelated to these changes, and more related to some stuff we need to fix on the backend to prevent `msftidy` from linting in ways that may be overly strict in some areas.

**Review by gwillcox-r7 — APPROVED** [review]
Changes LGTM. Awaiting approval from Alan due to him mentioning a possible merge conflict though.

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by gwillcox-r7** [comment]
Landing this now...

**Pull request merged**
